onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -407,7 +407,14 @@ Status CUDAExecutionProvider::EnqueueDeferredRelease() {
     cpu_buffers_info->buffers = buffers;
     // Release the ownership of cpu_buffers_info so that the underlying
     // object will keep alive until the end of ReleaseCpuBufferCallback.
-    CUDA_RETURN_IF_ERROR(cudaLaunchHostFunc(stream, ReleaseCpuBufferCallback, cpu_buffers_info.release()));
+    if (cpu_buffers_info->allocator->Info().alloc_type == OrtArenaAllocator) {
```

Comment:
This is the change to fix deferred release when CPU pinned memory is not allocated by BFCArena. For BFCArena cases, the `allocator->Free(ptr)` doesn't really call `cudaFreeHost`, so it's ok to call `allocator->Free(ptr)` in stream callback. Without BFCArena, `allocator->Free(ptr)` may invoke `cudaFreeHost` so we must call it directly from CPU.